group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -1211,13 +1192,99 @@ private CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> consumerGr
+        if (ownedTopicPartitions != null || memberEpoch == 0 || hasAssignedPartitionsChanged(member, updatedMember)) {
             response.setAssignment(createResponseAssignment(updatedMember));
         }
 
         return new CoordinatorResult<>(records, response);
     }
 
+    /**
+     * Reconciles the current assignment of the member if needed.
+     *
+     * @param groupId               The group id.
+     * @param member                The member to reconcile.
+     * @param currentPartitionEpoch The function returning the current epoch of
+     *                              a given partition.
+     * @param targetAssignmentEpoch The target assignment epoch.
+     * @param targetAssignment      The target assignment.
+     * @param ownedTopicPartitions  The list of partitions owned by the member. This
+     *                              is reported in the ConsumerGroupHeartbeat API and
+     *                              it could be null if not provided.
+     * @param records               The list to accumulate any new records.
+     * @return The received member if no changes have been made; or a new
+     *         member containing the new assignment.
+     */
+    private ConsumerGroupMember maybeReconcile(
+        String groupId,
+        ConsumerGroupMember member,
+        BiFunction<Uuid, Integer, Integer> currentPartitionEpoch,
+        int targetAssignmentEpoch,
+        Assignment targetAssignment,
+        List<ConsumerGroupHeartbeatRequestData.TopicPartitions> ownedTopicPartitions,
+        List<Record> records
+    ) {
+        if (member.isReconciledTo(targetAssignmentEpoch)) {
+            return member;
+        }
+
+        ConsumerGroupMember updatedMember = new CurrentAssignmentBuilder(member)
+            .withTargetAssignment(targetAssignmentEpoch, targetAssignment)
+            .withCurrentPartitionEpoch(currentPartitionEpoch)
+            .withOwnedTopicPartitions(ownedTopicPartitions)
+            .build();
+
+        if (!updatedMember.equals(member)) {
+            records.add(newCurrentAssignmentRecord(groupId, updatedMember));
+
+            log.info("[GroupId {}] Member {} new assignment state: epoch={}, previousEpoch={}, state={}, "
+                     + "assignedPartitions={} and revokedPartitions={}.",
+                groupId, updatedMember.memberId(), updatedMember.memberEpoch(), updatedMember.previousMemberEpoch(), updatedMember.state(),
+                formatAssignment(updatedMember.assignedPartitions()), formatAssignment(updatedMember.revokedPartitions()));
+
+            if (updatedMember.state() == MemberState.UNREVOKED_PARTITIONS) {
+                scheduleConsumerGroupRebalanceTimeout(
+                    groupId,
+                    updatedMember.memberId(),
+                    updatedMember.memberEpoch(),
+                    updatedMember.rebalanceTimeoutMs()
+                );
+            } else {
```

Comment:
The rebalance is not complete here because the member has to reach Stable for this. However, we only have a timeout on the revocation part. We basically want to member to not block revoked partitions forever.

Reply:
Yeah, that's right.

Reply:
i'm confused why we changed the name from revocation to rebalance timeout if it's actually the revocation timeout and a "rebalance" is considered complete only when we move to STABLE

Reply:
Did we come to a resolution here?

Reply:
So I use rebalance timeout in order to be consistent with the rebalanceTimeoutMs field. In the end, the name does not really matter here. The important part is how we apply it.

Reply:
I see. I think it is a little confusing for people trying to understand the algorithm, but I agree that it is important to have it apply and be used in a way that makes sense.

So just so I get it straight -- we have a timeout only for the revocation of the partitions. If we hit the timeout, the member is fenced and I assume we can count those partitions as revoked. (Or we no longer have to wait for them to be revoked)

At the point all partitions are revoked, we can try to assign new partitions to members. In this case, are we relying on the heartbeat to kick out members if they don't ack the new assignments?

Just trying to confirm the reason for the separate timeout here. Is it because revoking is more likely to fail even though the heartbeat still goes through, but not the same for assigning?

Reply:
> So just so I get it straight -- we have a timeout only for the revocation of the partitions. If we hit the timeout, the member is fenced and I assume we can count those partitions as revoked. (Or we no longer have to wait for them to be revoked)

Yep. This is correct. As soon as the member is fenced, all its partitions become available immediately.

> At the point all partitions are revoked, we can try to assign new partitions to members. In this case, are we relying on the heartbeat to kick out members out if they aren't responding? And we expect if requests are going through the assignment will occur?

Correct. We rely on the heartbeat to maintain the session of the member. If the member does not heartbeat at least once within the session timeout, we fence it.

> Just trying to confirm the reason for the separate timeout here. Is it because revoking is more likely to fail even though the heartbeat still goes through, but not the same for assigning?

The reason is that there actually two processes piggy backing on the heartbeat: 1) the session; and 2) the assignment reconciliation. The session is maintained by the regular heartbeat. The assignment reconciliation relies on the reported partitions provided by the member. It is totally possible for a member to heartbeat to maintain its sessions while being stuck on the assignment part. This is the main reason why we need the second timeout. We want to ensure that a member cannot block all the others forever if it does not releases its revoked partitions.

Regarding the assigning case, we don't apply the same timeout because we have discovered that the consumer could actually be stuck when a partition that does not exist anymore is assigned to it. In this case, the consumer won't be able to get the metadata for the newly assigned partition and it will retry forever. On the server, if we wait on the member to confirms that assignment, we cannot revoke the deleted partitions. This is the main reason why we treat both the revocation and the assignment differently. The revocation blocks the reconciliation process while the assignment does not.

Reply:
Ah. I was actually just investigating an incident where we had an issue with a deleted partition with the old protocol. 😅 

So this makes sense. I guess my only remaining question is whether there is a case where we could get stuck not consuming an existing partition that is assigned to a live member, but that member is not able to get the assignment but is able to heartbeat. This is the case where an assignment timeout could also free this assigned partition to be assigned to another member that may be able to take it. 

I agree though that it is better in the current state (not blocking the members from getting to the reconcillation etc).

Reply:
Yeah, this could potentially happen. That's the tradeoff that we have to make here. However, I would like to point out that on the client, the partition is considered assigned even if the client may not be able to fetch from it (e.g. in case of stale metadata) so the contract is loose anyway here.

If it turns out to be an issue in the future, I think that we could have a timer for individual partitions. I refrained myself from doing this because it brings a lot of complexity and it does not seem worth it at the moment.